docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+  a built-in consensus system in the pageserver, although this could in principle
+  be changed without changing the storage format.
+- **The safekeeper may refuse RPCs from zombie pageservers** via the node generation number.
+- **Deletions are deferred via deletion lists** with occasional synchronization
+  to the control plane to ensure safety.
+
+### Generation numbers
+
+Two logical entities will get new "generation numbers", which are monotonically increasing
+integers with a global guarantee that they will not be re-used:
+
+- Pageservers: a generation ID is acquired at startup and lasts until the process ends or another
+  process with the same node ID starts.
+- Tenant attachments: for a given tenant, each time its attachment is changed, a per-tenant generation
+  number increases.
+
+This provides some important invariants:
+
+- If there are multiple pageservers running with the same node ID, they are guaranteed to have
+  a different generation ID.
+- If there are multiple pageservers running with the same node ID, we may unambiguously know which
+  of them "wins" by picking the higher generation number.
+- If two pageservers have the same tenant attached, they are guaranteed to have different attachment
+  generation numbers.
+
+The node generation number defines a global "in" definition for pageserver nodes: a node whose
+generation matches the generation the control plane most recently issued is a member of the cluster
+in good standing. Any node with any older generation ID is not.
+
+#### Why use per-tenant _and_ per-node generation numbers?
+
+The essential generation number is the tenant attachment number: this alone would be sufficient
+to implement safe migration and failover. However, it would also require that all such events
+carry an O(tenant_count) cost, or to have a global sequence number for all changes to attachments (which
+would imply synchronization between all attachment changes): the pageserver generation number
+is included to avoid writing this kind of system design constraint into the storage format.
+
+The pageserver generation number acts as an optimization and also provides flexibility in future designs:
+
+- if we wanted to have pageservers do their own HA in future without control plane in the loop on a per-tenant basis,
+  then node generation numbers make that possible.
+- the pageserver generation number also enables future storage of per-pageserver data in S3 (e.g.
+  housekeeping type content) that may not be within the scope of a tenant attachment.
+- pageserver generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+The tenant attachment generation and node generation are mostly independent, but have one key
+coupling:
+
+- Control plane must not "steal" a tenant from a node (by attaching to another node without detaching
```

Comment:
This text is gone, and replaced with:

>The two generation numbers have a different behavior for stale generations:
> - A pageserver with a stale generation number should immediately terminate: it is never intended
>   to have two pageservers with the same node ID.
> - An attachment with a stale generation number is permitted to continue operating (ingesting WAL
>  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
>  facilitating live migration (this will be articulated in the next RFC that describes HA and
>  migrations).

So we will permit multi-attaching tenants (seeing a newer generation doesn't fence the old one) during migrations, but in this multi-attached state deletions are paused.

Reply:
896bfdb2299d2a4f4b1b942a35f5704ce801d020